Feature/map model

FellonOne opened the pull request:
release map model and map drawer;
extremely many changes;

FellonOne committed:
test-map store structure

FellonOne committed:
Merge branch 'develop' into feature/map-model

FellonOne committed:
Merge branch 'develop' into feature/map-model

FellonOne committed:
-only-for-level-geneartor-page

FellonOne committed:
temp-commit-old

FellonOne committed:
Merge branch 'develop' into feature/map-model

FellonOne committed:
pre-release map;

FellonOne committed:
old commit for backup;

FellonOne committed:
pre-sunday commit;

FellonOne committed:
copy from work pc;

FellonOne committed:
pre-final map generator;

FellonOne committed:
add final levelGenerator function;

FellonOne committed:
fix issue in LevelGenerator;

FellonOne committed:
Merge branch 'develop' into feature/map-model
fix issues

kdeit commented on app/src/components/Menu/Menu.tsx:
```diff
@@ -4,18 +4,21 @@ import {Link} from 'react-router-dom';
 const mapDispatch = {
-    signOut: signOutAction
+    signOut: signOutAction,
+    generateLevel: levelGenerateAction
```
Похоже на артефакт старого кода

kdeit commented on app/src/pages/StartPage/StartPage.tsx:
```diff
@@ -30,14 +31,15 @@ const mapState = (state: AppStoreState) => ({
     charSetName: gameCharSetNameAction,
-    gameSetState: gameSetStateAction
+    gameSetState: gameSetStateAction,
+    generateMap: levelRandomGenerateAction
```
А вот тут может логику запуска генератора карты перенести в сагу и пускать когда меняется стейт игры? И всё в одном месте будет.

FellonOne replied:
Да, это чисто ради демонстрации 
Надо будет перенести в interlude

kdeit commented on app/src/store/Game/sagas.ts:
```diff
@@ -21,6 +23,11 @@ function* gameSetStateWorker(action: GameSetStateAction) {
             break;
         }
         case 'GAME': {
+            const state: AppStoreState = yield select();
```
Тут можно взять часть state'а

kdeit commented on app/src/services/LevelGenerator/LevelGenerator.ts:
```diff
@@ -0,0 +1,444 @@
+import cloneDepp from 'lodash/cloneDeep';
```
Ошибка в названии :) cloneDeep

kdeit commented on app/src/pages/LevelGenerator/LevelGenerator.tsx:
```diff
@@ -0,0 +1,73 @@
+
+            <div className="mt-5">
+                <Button
+                    onClick={() => {
```
Очень сложно воспринимается. Может в отдельную функцию?

FellonOne replied:
Да, это надо будет вообще удалить (файл LevelGenerator и его упоминания)

kdeit commented on app/src/pages/LevelGenerator/LevelGenerator.tsx:
```diff
@@ -0,0 +1,73 @@
+                    row.map((cell, j) => (
+                        <input
+                            className="generator-grid__cell"
+                            key={`${cell}_${Date.now()}_${Math.round(Math.random() * 99999)}_${Date.now()}_${Math.round(
```
Выглядит устрашающе :)

kdeit reviewed (approved):
Я бы заапрувил и пошёл дальше. Критичного нет. А мелочи отрефакторить можно в процессе.

FellonOne committed:
fix issues from PR;
decouple LevelGenerator logic;

art4vm commented on app/src/core/sprites/map/types.ts:
```diff
@@ -0,0 +1,24 @@
+import {Sprite} from '../types';
+
+export interface Wall {
+    CORNER_TL: Sprite;
+    CORNER_TR: Sprite;
+    CORNER_BL: Sprite;
+    CORNER_BR: Sprite;
+    TOP: Sprite;
+    SIDE: Sprite;
+    LEDGE: Sprite;
+    CROSS_TOP: Sprite;
+    CROSS_BOTTOM: Sprite;
+    CROSS_LEFT: Sprite;
+    CROSS_RIGHT: Sprite;
+    CROSS: Sprite;
+    FILL: Sprite;
+}
+
+export interface Floor {
+    ROOM: Sprite;
+    SINGLE: Sprite;
+    VERTICAL: Sprite;
+    HORIZONTAL: Sprite;
+}
```
В таких типах нет необходимости, они дублируют сам объект

FellonOne replied:
Они используются в коде
В месте:
`export type BackgroundAsset = {part: 'FLOOR' | 'WALL'; type: keyof Floor | keyof Wall};`

art4vm replied:
Я видел где это используется
```ts
export type BackgroundAsset = {type: 'FLOOR'; part: keyof typeof floorSprites} | {type: 'WALL'; part: keyof typeof wallSprites}
```

art4vm commented on app/src/store/Game/reducer.ts:
```diff
@@ -2,6 +2,7 @@ import cloneDeep from 'lodash/cloneDeep';
 
 import {GAME_ACTION_TYPES, GameState, GameActions} from './types';
 
+// TODO mock
```
это не мок

FellonOne replied:
да, видимо при мердже косяк(

art4vm commented on app/src/router/AppRouter/AppRouter.tsx:
```diff
@@ -14,9 +14,6 @@ import {
 
 import {PrivateRoute} from '../PrivateRoute/PrivateRoute';
 
-/**
- * Роутер
- */
```
зачем это удаление в этом пр?

FellonOne replied:
Видимо при мердже затерлось

art4vm commented on app/src/core/main.canvas.ts:
```diff
@@ -30,6 +30,25 @@ export function setMainCanvas(canvasElement: HTMLCanvasElement | null): void {
     Promise.all([characterSprites]).then(([CHAR]) => {
         character = CHAR.PALADIN;
         drawImage(fleft, ftop, character);
+        window.addEventListener('keydown', (e) => {
+            ctx.clearRect(fleft * GRID, ftop * GRID, character.width, character.height);
+            switch (e.key) {
+                case 'ArrowLeft':
+                    movef(-1, 0);
+                    break;
+                case 'ArrowRight':
+                    movef(1, 0);
+                    break;
+                case 'ArrowUp':
+                    movef(0, -1);
+                    break;
+                case 'ArrowDown':
+                    movef(0, 1);
+                    break;
+                default:
+            }
+            drawImage(fleft, ftop, character);
+        });
```
Это было вынесено в engine, похоже на код от старого патча

FellonOne replied:
Тоже самое, мердж

art4vm commented on app/src/core/bg.canvas.ts:
```diff
@@ -1,38 +1,38 @@
 
-import type {Sprite} from './types';
+import type {Sprite} from './sprites/types';
+import {BackgroundMap} from '~/store/Level/types';
```
type

art4vm commented on app/src/core/types.ts:
```diff
@@ -15,15 +15,6 @@ export interface Sprite {
     height: number;
 }
```
Если Sprite был вынесен, наверное тут его нужно удалить

art4vm commented on app/src/core/main.canvas.ts:
```diff
@@ -49,9 +68,7 @@ function drawImage(x: number, y: number, sprite: Sprite) {
     fleft = (fleft + x + width) % width;
     ftop = (ftop + y + height) % height;
-    drawImage(fleft, ftop, character);
 }
```
очистка была сюда перенесена не просто так, кажется это остаток от какого-то старого кода

art4vm commented on app/src/core/bg.canvas.ts:
```diff
@@ -1,38 +1,38 @@
+
 import {setCanvas} from './engine';
-import {floorSprites} from './sprites/floor';
+import {wallSprites} from './sprites/map/wall';
```
в пр много изменений связанных с изменением вложенности, предлагаю вернуть как было, тк смысловой нагрузки в этом нет

art4vm commented on app/src/core/sprites/map/wall.ts:
```diff
@@ -10,7 +10,7 @@ assetMap.src = './assets/wall.png';
  * <image, сдвиг в карте по X, сдвиг в карте по Y, ширина, высота>
  */
-const SPRITES = {
+const SPRITES: AssetMap = {
```
тип тут не нужен

art4vm commented on app/src/services/LevelGenerator/__levels.ts:
```diff
@@ -0,0 +1,72 @@
+import {BackgroundMap} from '../../store/Level/types';
+
+export interface DefaultLevel {
+    startPoint: [number, number];
+    endPoint: [number, number];
+    map: number[][];
+    completeMap?: BackgroundMap;
+}
```
смущает название. А не default как-то иначе будет выглядеть? Так же тип стоит вынести в файл с типами

FellonOne replied:
Это мок хранения уровней

art4vm commented on app/src/store/Level/actions.ts:
```diff
@@ -0,0 +1,23 @@
+        type: LEVEL_ACTION_TYPES.GENERATE_LEVELS,
+        payload: generatorConfig
+    };
+}
```
2 экшена с одинаковыми типами, кажется так не стоит делать

art4vm commented on app/src/store/Level/actions.ts:
```diff
@@ -0,0 +1,23 @@
+import type {Level, LevelGenerateAction, LevelSaveAction} from './types';
+import {LEVEL_ACTION_TYPES} from './types';
```
тут импорт из одного и того же файла, нужно объединить

FellonOne replied:
В одном случае - импортируем только типы
Во втором - Enum объект

art4vm commented on app/src/store/Level/actions.ts:
```diff
@@ -0,0 +1,23 @@
+import type {Level, LevelGenerateAction, LevelSaveAction} from './types';
+import {LEVEL_ACTION_TYPES} from './types';
+import {GeneratorConfiguration} from '~/services/LevelGenerator/types';
```
type

art4vm commented on app/src/store/Level/sagas.ts:
```diff
@@ -0,0 +1,28 @@
+import {ForkEffect, put, takeLeading} from 'redux-saga/effects';
+
+import {LevelGenerator} from '../../services/LevelGenerator/LevelGenerator';
```
```suggestion
import {LevelGenerator} from '~/services/LevelGenerator/LevelGenerator';
```

art4vm commented on app/src/store/Level/types.ts:
```diff
@@ -0,0 +1,60 @@
+import {Floor, Wall} from '../../core/sprites/map';
```
```suggestion
import {Floor, Wall} from '~/core/sprites/map';
```

art4vm commented on app/src/store/Level/types.ts:
```diff
@@ -0,0 +1,60 @@
+import {Floor, Wall} from '../../core/sprites/map';
+
+import {GeneratorConfiguration} from '../../services/LevelGenerator/types';
```
```suggestion
import {GeneratorConfiguration} from '~/services/LevelGenerator/types';
```

art4vm commented on app/src/store/types.ts:
```diff
@@ -1,12 +1,14 @@
 import type {LeadeboardState} from './Leaderboard/types';
-import type {ToasterState} from './Toaster/types';
+import {LevelState} from './Level/types';
+import {ToasterState} from './Toaster/types';
```
```suggestion
import type {LevelState} from './Level/types';
import type {ToasterState} from './Toaster/types';
```

art4vm commented on app/src/store/Level/types.ts:
```diff
@@ -0,0 +1,60 @@
+
+export type BackgroundAsset = {part: 'FLOOR' | 'WALL'; type: keyof Floor | keyof Wall};
+export interface BackgroundPart {
+    asset: BackgroundAsset | null;
```
```suggestion
    asset?: BackgroundAsset;
```

art4vm commented on app/src/services/LevelGenerator/__levels.ts:
```diff
@@ -0,0 +1,72 @@
+import {BackgroundMap} from '../../store/Level/types';
```
```suggestion
import {BackgroundMap} from '~/store/Level/types';
```

art4vm commented on app/src/services/LevelGenerator/utils.ts:
```diff
@@ -0,0 +1,57 @@
+    return true;
+};
+
+export const randSort = (array: number[][]): number[][] => {
```
тут array это просто одномерный массив

art4vm commented on app/src/core/sprites/map/wall.ts:
```diff
@@ -43,6 +43,22 @@ const VARIANTS: AssetVariants = {
     BRICK_4: {
         x: 0,
         y: 12
+    },
+    BRICK_5: {
+        x: 0,
+        y: 15
+    },
+    BRICK_6: {
+        x: 0,
+        y: 18
+    },
+    BRICK_7: {
+        x: 0,
+        y: 21
+    },
+    BRICK_8: {
+        x: 0,
+        y: 24
```
В этом нет смысла, тк не используется

art4vm commented on app/src/services/LevelGenerator/LevelGenerator.ts:
```diff
@@ -0,0 +1,388 @@
+const MAX_SUM = 9;
+
+function mapWall(inputLevel: number[][]): number[][] {
+    const level = cloneDeep(inputLevel);
```
Тут нет смысла копировать, тк весь массив все равно перезаписывается. Так же сложно читать код, когда то inputLevel то level используется. В циклах и проверках можно использовать inputLevel

FellonOne replied:
Нет, к сожалению нужно
Для построения стен используется первичная информация об уровне 0 и 1
Т.е. если сразу делать внутри inputLevel, то там будут уже изменения и алгоритм не будет работать

А вот для дорог внутри замка нужна инфа по уже построенным дорогам
Чтоб не строить лишнии

art4vm commented on app/src/services/LevelGenerator/LevelGenerator.ts:
```diff
@@ -0,0 +1,388 @@
+                        level[k][l] = -1;
+                    });
+
+                    level[i - 1][j - 1] = MapPartials.CORRIDOR_ROAD_3_3;
```
Вместо
```
MapPartials.CORRIDOR_ROAD_3_3
```
думаю можно сразу ставить 
```
{
    part: 'FLOOR',
    type: 'SINGLE'
}
```

сразу сократиться куча кода а так же необходимость в 2 циклах

FellonOne replied:
это расширение для будущих изменений
тип, на MapPartials структуре раскрашивать уровни в разные цвета, заполнения всякими фичами и тд

art4vm commented on app/src/services/LevelGenerator/__levels.ts:
```diff
@@ -0,0 +1,72 @@
+    startPoint: [1, 0],
+    endPoint: [22, 30],
+    map: [
+        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
```
Предлагаю как-то решить проблему с крайними полями, тк выглядит костыльно да и игровой поле сокращается

FellonOne replied:
Почему выглядит костыльно? Красивый периметр получается
Можно оставить точку *входа* в локацию, оставив дыру в периметре и все

FellonOne committed:
fix issues in PR;

art4vm commented on app/src/store/Game/sagas.ts:
```diff
@@ -17,10 +20,15 @@ function* gameSetStateWorker(action: GameSetStateAction) {
+            const state: AppStoreState = yield select();
+            const {map} = state.level.levels[state.game.level.number ?? 0];
+
+            yield call(drawMap.bind(null, map));
```
```suggestion
            yield call(drawMap, map);
```

art4vm committed:
Добавлены новые спрайты и слои для canvas

art4vm committed:
Изменения в типизации карт спрайтов

Добавлена обработка карты спрайтов с объектами
Добавлены типизации для карт спрайтов
Добавлены трансформеры для преобразования спрайтов в строковые ассеты и обратно
Исправлена потеря типов в variantFactory

Дополнительно: общие параметры окружения игры вынесены в отдельный файл
 params.ts

FellonOne committed:
fix issues from PR v2;

art4vm committed:
Добавлена обертка для типизации Object.entries

art4vm committed:
Правка путей до params и исправление типов

art4vm committed:
Правка путей до params и исправление типов. Fix

FellonOne committed:
add objects and chars implementation;
add new level (by Artyom);
update render function;

art4vm commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Реализация рендера карты на слое canvasBG](https://trello.com/c/xJyGvupe/61-%D1%80%D0%B5%D0%B0%D0%BB%D0%B8%D0%B7%D0%B0%D1%86%D0%B8%D1%8F-%D1%80%D0%B5%D0%BD%D0%B4%D0%B5%D1%80%D0%B0-%D0%BA%D0%B0%D1%80%D1%82%D1%8B-%D0%BD%D0%B0-%D1%81%D0%BB%D0%BE%D0%B5-canvasbg)
